File: daily_challenge_leetcode/3346.py
# Tc = O(maxEl)
# Sc = O(maxEl)


from typing import List

class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        maxEl = max(nums) + k
        freq = [0] * (maxEl + 1)
        for num in nums:
            freq[num] += 1
        
        # cumulative sum of freq
        for i in range(1, maxEl + 1):
            freq[i] += freq[i - 1]
        
        result = 0
        for target in range(maxEl + 1):
            if freq[target] == 0:
                continue
            leftNum = max(0, target - k)
            rightNum = min(maxEl, target + k)

            totalCount = freq[rightNum] - (freq[leftNum - 1] if leftNum > 0 else 0)
            targetCount = freq[target] - (freq[target - 1] if target > 0 else 0)

            needConversion = totalCount - targetCount
            maxPossibleFreq = targetCount + min(needConversion, numOperations)
            result = max(result, maxPossibleFreq)
        
        return result

File: daily_challenge_leetcode/3346.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} numOperations
 * @return {number}
 */
var maxFrequency = function (nums, k, numOperations) {
    const maxEl = Math.max(...nums) + k;
    const freq = new Array(maxEl + 1).fill(0);
    for (const num of nums) {
        freq[num]++;
    }

    for (let i = 1; i <= maxEl; i++) {
        freq[i] += freq[i - 1];
    }

    let result = 0;
    for (let target = 0; target <= maxEl; target++) {
        if (freq[target] === 0) {
            continue;
        }
        const leftNum = Math.max(0, target - k);
        const rightNum = Math.min(maxEl, target + k);

        const totalCount = freq[rightNum] - (leftNum > 0 ? freq[leftNum - 1] : 0);
        const targetCount = freq[target] - (target > 0 ? freq[target - 1] : 0);
        const needConversion = totalCount - targetCount;

        const maxPossibleFreq = targetCount + Math.min(needConversion, numOperations);
        result = Math.max(result, maxPossibleFreq);
    }

    return result;
};